File: starters/dotnet/FProto.Core/Session.cs
namespace FProto.Core;

public enum SessionStatus
{
    Connecting,
    Active,
    Sleeping,
    Expired
}

public enum SessionEvent
{
    HandshakeOK,
    HandshakeFail,
    PacketReceived,
    IdleTimeout,
    ExpiryTimeout,
    ExplicitLogout
}

public record TransitionResult(SessionStatus Next, bool Destroyed);

public class InvalidTransitionException : Exception
{
    public SessionStatus Status { get; }
    public SessionEvent Event { get; }

    public InvalidTransitionException(SessionStatus status, SessionEvent ev)
        : base($"invalid transition: {status} + {ev}")
    {
        Status = status;
        Event = ev;
    }
}

public static class SessionMachine
{
    private static readonly Dictionary<(SessionStatus, SessionEvent), TransitionResult> Transitions = new()
    {
        [(SessionStatus.Connecting, SessionEvent.HandshakeOK)] = new(SessionStatus.Active, false),
        [(SessionStatus.Connecting, SessionEvent.HandshakeFail)] = new(SessionStatus.Expired, true),
        [(SessionStatus.Active, SessionEvent.PacketReceived)] = new(SessionStatus.Active, false),
        [(SessionStatus.Active, SessionEvent.IdleTimeout)] = new(SessionStatus.Sleeping, false),
        [(SessionStatus.Active, SessionEvent.ExplicitLogout)] = new(SessionStatus.Expired, false),
        [(SessionStatus.Sleeping, SessionEvent.PacketReceived)] = new(SessionStatus.Active, false),
        [(SessionStatus.Sleeping, SessionEvent.ExpiryTimeout)] = new(SessionStatus.Expired, false),
    };

    public static TransitionResult Transition(SessionStatus current, SessionEvent ev)
    {
        if (Transitions.TryGetValue((current, ev), out var result))
            return result;
        throw new InvalidTransitionException(current, ev);
    }
}

public record SessionConfig(
    int IdleTimeoutSecs = 300,
    int ExpiryTimeoutSecs = 86400,
    int HandshakeTimeoutSecs = 30,
    int MaxConnections = 10000
);

File: starters/node/src/session.ts
export enum Status {
  Connecting = 'connecting',
  Active = 'active',
  Sleeping = 'sleeping',
  Expired = 'expired',
}

export enum Event {
  HandshakeOK = 'handshake_ok',
  HandshakeFail = 'handshake_fail',
  PacketReceived = 'packet_received',
  IdleTimeout = 'idle_timeout',
  ExpiryTimeout = 'expiry_timeout',
  ExplicitLogout = 'explicit_logout',
}

export interface TransitionResult {
  next: Status;
  destroyed: boolean;
}

export class InvalidTransitionError extends Error {
  constructor(
    public status: Status,
    public event: Event,
  ) {
    super(`invalid transition: ${status} + ${event}`);
    this.name = 'InvalidTransitionError';
  }
}

type TransitionKey = `${Status}:${Event}`;

const transitions = new Map<TransitionKey, TransitionResult>([
  [`${Status.Connecting}:${Event.HandshakeOK}`, { next: Status.Active, destroyed: false }],
  [`${Status.Connecting}:${Event.HandshakeFail}`, { next: Status.Expired, destroyed: true }],
  [`${Status.Active}:${Event.PacketReceived}`, { next: Status.Active, destroyed: false }],
  [`${Status.Active}:${Event.IdleTimeout}`, { next: Status.Sleeping, destroyed: false }],
  [`${Status.Active}:${Event.ExplicitLogout}`, { next: Status.Expired, destroyed: false }],
  [`${Status.Sleeping}:${Event.PacketReceived}`, { next: Status.Active, destroyed: false }],
  [`${Status.Sleeping}:${Event.ExpiryTimeout}`, { next: Status.Expired, destroyed: false }],
]);

export function transition(current: Status, event: Event): TransitionResult {
  const key: TransitionKey = `${current}:${event}`;
  const result = transitions.get(key);
  if (!result) {
    throw new InvalidTransitionError(current, event);
  }
  return { ...result };
}

export interface SessionConfig {
  idleTimeoutSecs: number;
  expiryTimeoutSecs: number;
  handshakeTimeoutSecs: number;
  maxConnections: number;
}

export function defaultConfig(): SessionConfig {
  return {
    idleTimeoutSecs: 300,
    expiryTimeoutSecs: 86400,
    handshakeTimeoutSecs: 30,
    maxConnections: 10000,
  };
}
